clang-tools-extra/clang-tidy/modernize/UseNodiscardCheck.cpp

```diff
@@ -84,33 +82,32 @@ void UseNodiscardCheck::storeOptions(ClangTidyOptions::OptionMap &Opts) {
 }
 
 void UseNodiscardCheck::registerMatchers(MatchFinder *Finder) {
-  auto FunctionObj =
-      cxxRecordDecl(hasAnyName("::std::function", "::boost::function"));
-
   // Find all non-void const methods which have not already been marked to
   // warn on unused result.
   Finder->addMatcher(
       cxxMethodDecl(
-          allOf(isConst(), isDefinitionOrInline(),
-                unless(anyOf(
-                    returns(voidType()),
-                    returns(hasDeclaration(decl(hasAttr(clang::attr::WarnUnusedResult)))),
-                    isNoReturn(), isOverloadedOperator(),
-                    isVariadic(), hasTemplateReturnType(),
-                    hasClassMutableFields(), isConversionOperator(),
-                    hasAttr(clang::attr::WarnUnusedResult),
-                    hasType(isInstantiationDependentType()),
-                    hasAnyParameter(anyOf(
-                        parmVarDecl(anyOf(hasType(FunctionObj),
-                                          hasType(references(FunctionObj)))),
-                        hasType(isNonConstReferenceOrPointer()),
-                        hasParameterPack()))))))
+          allOf(isDefinitionOrInline(),
+                unless(anyOf(isLambda(), returns(voidType()), isNoReturn(),
```

Comment:
Where does the exception for lambdas come from? It's not mentioned in CS.R.5

Reply:
It should be mentioned. There is no way to add nodiscard to a lambda `operator()`

Reply:
Not with current standard. C++2b  allows `[][[nodiscard]]()->int { return 2; }`